### PR TITLE
added some category theory into FunctionK document

### PR DESCRIPTION
**Pull request by kailuowang**
I think this will be helpful because it's often referred as Natural ~~Transformer~~ Transformation in other context.

### CODE REVIEW

**Comment by wedens** [comment]
I think it's known as "natural transformation", not "transformer" :)

**Comment by kailuowang** [comment]
@wedens thanks for pointing out. Corrected.

File: docs/src/main/tut/datatypes/functionk.md [hunk 1]
```diff
@@ -121,3 +121,25 @@ type ErrorOr[A] = Either[String, A]
 val errorOrFirst: FunctionK[List, ErrorOr] =
   λ[FunctionK[List, ErrorOr]](_.headOption.toRight("ERROR: the list was empty!"))
 ```
+
+## Natural Transformation
+
+In category theory, a [Natural Transformation](https://en.wikipedia.org/wiki/Natural_transformation) provides a morphism between Functors while preserving the internal structures. It's one of the most fundamental notions of category theory.
```
Comment (SystemFw) [inline docs/src/main/tut/datatypes/functionk.md]: > while preserving the internal structures

I think _the internal structure_ is slightly better. It refers to the composition of morphisms (similarly to what a functor does)

File: docs/src/main/tut/datatypes/functionk.md [hunk 2]
```diff
@@ -121,3 +121,25 @@ type ErrorOr[A] = Either[String, A]
...
+(fa:F[A]) => fK(F.map(fa)(f))
+```
+and
+```Scala
+(fa: F[A]) => G.map(fK(fa))(f)
+````
+
+We don't need to write a law to test the implementation of the `fk` for the above to be true. It's automatically given by its parametric polymorphism.
+
+This is why a parametric polymorphic function `FunctionK[F, G]` is sometime referred as a Natural Transformation.
```
Comment (SystemFw) [inline docs/src/main/tut/datatypes/functionk.md]: Pedantry warning (sorry): it should be `sometimes`, not `sometime`.  :)

File: docs/src/main/tut/datatypes/functionk.md [hunk 3]
```diff
@@ -121,3 +121,25 @@ type ErrorOr[A] = Either[String, A]
...
+However, they are two different concepts. `FunctionK` is a stronger and stricter construction than Natural Transformation.
+
```
Comment (SystemFw) [inline docs/src/main/tut/datatypes/functionk.md]: I'm not entirely sure this is worth including unless we can give an explanation as to how do they differ

File: docs/src/main/tut/datatypes/functionk.md [hunk 4]
```diff
@@ -121,3 +121,25 @@ type ErrorOr[A] = Either[String, A]
...
   λ[FunctionK[List, ErrorOr]](_.headOption.toRight("ERROR: the list was empty!"))
 ```
+
+## Natural Transformation
+
+In category theory, a [Natural Transformation](https://en.wikipedia.org/wiki/Natural_transformation) provides a morphism between Functors while preserving the internal structures. It's one of the most fundamental notions of category theory.
+
+If we have two `Functor`s: `F` and `G`, then, being parametric polymorphic, `FunctionK[F, G]` is automatically a Natural Transformation between them. That is, if we have a `fk: F ~> G`, then for any combination of `A`, `B` and function `f: A => B`, the following two functions are equivalent:
+```Scala
+(fa:F[A]) => fK(F.map(fa)(f))
```
Comment (alexknvl) [inline docs/src/main/tut/datatypes/functionk.md]: Minor inconsistency: `fK` and `fk`.

**Comment by kailuowang** [comment]
@ceedubs I am interested in hearing your thoughts on this.

File: docs/src/main/tut/datatypes/functionk.md [hunk 5]
```diff
@@ -121,3 +121,24 @@ type ErrorOr[A] = Either[String, A]
...
   λ[FunctionK[List, ErrorOr]](_.headOption.toRight("ERROR: the list was empty!"))
 ```
+
+## Natural Transformation
+
+In category theory, a [Natural Transformation](https://en.wikipedia.org/wiki/Natural_transformation) provides a morphism between Functors while preserving the internal structure. It's one of the most fundamental notions of category theory.
+
+If we have two `Functor`s: `F` and `G`, then, being parametric polymorphic, `FunctionK[F, G]` is automatically a Natural Transformation between them. That is, if we have a `fk: F ~> G`, then for any combination of `A`, `B` and function `f: A => B`, the following two functions are equivalent:
+```Scala
+(fa:F[A]) => fK(F.map(fa)(f))
```
Comment (ChristopherDavenport) [inline docs/src/main/tut/datatypes/functionk.md]: This has `fK` where is is defined above as `fk`

File: docs/src/main/tut/datatypes/functionk.md [hunk 6]
```diff
@@ -121,3 +121,24 @@ type ErrorOr[A] = Either[String, A]
...
+(fa:F[A]) => fk(F.map(fa)(f))
+```
+and
+```Scala
+(fa: F[A]) => G.map(fk(fa))(f)
+````
+
+We don't need to write a law to test the implementation of the `fk` for the above to be true. It's automatically given by its parametric polymorphism.
+
+This is why a parametric polymorphic function `FunctionK[F, G]` is sometimes referred as a Natural Transformation. However, they are two different concepts.
```
Comment (ceedubs) [inline docs/src/main/tut/datatypes/functionk.md]: Minor: is there a particular reason that you capitalized "Natural Transformation" here?

**Comment by ceedubs** [comment]
Thanks @kailuowang. I really like linking this back to the category theory (and probably also providing some insight to people who are used to the `NaturalTransformation` name in scalaz). I also like that it's added as a separate section at the end so people can get into the theory if they want to but can stop short of it if they prefer. 👍

**Comment by ceedubs** [comment]
@adelbertc as someone who has written a lot of docs and read up on your category theory, does this look good to you?

File: docs/src/main/tut/datatypes/functionk.md [hunk 7]
```diff
@@ -121,3 +121,24 @@ type ErrorOr[A] = Either[String, A]
 val errorOrFirst: FunctionK[List, ErrorOr] =
   λ[FunctionK[List, ErrorOr]](_.headOption.toRight("ERROR: the list was empty!"))
 ```
+
+## Natural Transformation
+
+In category theory, a [Natural Transformation](https://en.wikipedia.org/wiki/Natural_transformation) provides a morphism between Functors while preserving the internal structure. It's one of the most fundamental notions of category theory.
```
Comment (adelbertc) [inline docs/src/main/tut/datatypes/functionk.md]: ncatlab might be a better reference than Wikipedia for such a subject :-) https://ncatlab.org/nlab/show/natural+transformation

File: docs/src/main/tut/datatypes/functionk.md [hunk 8]
```diff
@@ -121,3 +121,24 @@ type ErrorOr[A] = Either[String, A]
...
+
+If we have two `Functor`s: `F` and `G`, then, being parametric polymorphic, `FunctionK[F, G]` is automatically a Natural Transformation between them. That is, if we have a `fk: F ~> G`, then for any combination of `A`, `B` and function `f: A => B`, the following two functions are equivalent:
```
Comment (adelbertc) [inline docs/src/main/tut/datatypes/functionk.md]: Perhaps drop the colon, and use 'parametricity', e.g.

If we have two `Functor`s `F` and `G`, `FunctionK[F, G]` is a natural transformation via parametricity. That is, given `fk: FunctionK[F, G]`, for all functions `A => B` the following are equivalent:

File: docs/src/main/tut/datatypes/functionk.md [hunk 9]
```diff
@@ -121,3 +121,24 @@ type ErrorOr[A] = Either[String, A]
...
+```Scala
+(fa:F[A]) => fk(F.map(fa)(f))
+```
+and
+```Scala
+(fa: F[A]) => G.map(fk(fa))(f)
+````
+
+We don't need to write a law to test the implementation of the `fk` for the above to be true. It's automatically given by its parametric polymorphism.
```
Comment (adelbertc) [inline docs/src/main/tut/datatypes/functionk.md]: by parametricity

File: docs/src/main/tut/datatypes/functionk.md (hunk 6 above)
Comment (adelbertc) [inline docs/src/main/tut/datatypes/functionk.md]: Probably don't need to capitalize

Comment (adelbertc) [inline docs/src/main/tut/datatypes/functionk.md]: This line could use some elaboration.. maybe say something about how `FunctionK` is more powerful than natural transformations, or natural transformations can be implemented in terms of `FunctionK` (I prefer the latter)

**Comment by adelbertc** [comment]
@ceedubs you flatter me :-)

File: docs/src/main/tut/datatypes/functionk.md (hunk 6 above)
Comment (kailuowang) [inline docs/src/main/tut/datatypes/functionk.md]: We had a discussion about this on gitter, and @edmundnoble pointed out that natural transformation could be deemed as more powerful in some ways. My intuition is that `FunctionK` is more powerful in one category but natural transformation is a concept that could involve two categories. Maybe when we restrict natural transformation to endofunctors we can say something about it. In conclusion we feel it's too complex a thing to include here.

**Pull request merged**